### PR TITLE
Initial implementation of SMIOL file open/close routines

### PR DESCRIPTION
**Pull request by mgduda**
This merge provides initial implementations of the `SMIOL_open_file` and `SMIOL_close_file`
routines using the parallel-netCDF library, and it also provides unit tests for these routines.

As of this merge, the `SMIOL_open_file` routine can only create a new file -- it is not yet
possible to open existing files.

### CODE REVIEW

**Comment by mgduda** [comment]
@MiCurry For this PR, I'd suggest we use a merge commit rather than rebasing and fast-forwarding.

File: src/smiol.c [hunk 1]
```diff
@@ -135,11 +139,51 @@ int SMIOL_inquire(void)
  *
  * Opens a file within a SMIOL context.
  *
- * Detailed description.
+ * Creates or opens the file specified by filename within the provided SMIOL
+ * context.
+ *
+ * Upon successful completion, SMIOL_SUCCESS is returned, and the file handle argument
+ * will point to a valid file handle. Otherwise, the file handle is NULL and an error
+ * code other than SMIOL_success is returned.
```
Comment (MiCurry) [inline src/smiol.c]: 'SMIOL_success' should be 'SMIOL_SUCCESS'

File: src/smiolf.F90 [hunk 2]
```diff
@@ -189,14 +192,72 @@ end function SMIOLf_inquire
     !
     !> \brief Opens a file within a SMIOL context
     !> \details
-    !>  Detailed description of what this routine does.
+    !>  Creates or opens the file specified by filename within the provided SMIOL
+    !>  context.
+    !>
+    !>  Upon successful completion, SMIOL_SUCCESS is returned, and the file handle argument
+    !>  will point to a valid file handle. Otherwise, the file handle is not associated
+    !>  and an error code other than SMIOL_success is returned.
```
Comment (MiCurry) [inline src/smiolf.F90]: 'SMIOL_success' should also be 'SMIOL_SUCCESS' here

File: src/smiol.c [hunk 3]
```diff
@@ -135,11 +139,51 @@ int SMIOL_inquire(void)
...
+	 * Save pointer to context for this file
+	 */
+	(*file)->context = context;
+
+#ifdef SMIOL_PNETCDF
+	if ((ierr = ncmpi_create(MPI_Comm_f2c(context->fcomm), filename, NC_NOCLOBBER, MPI_INFO_NULL, &((*file)->ncidp))) != NC_NOERR) {
+		free((*file));
+		(*file) = NULL;
+		return -996;
+	}
```
Comment (MiCurry) [inline src/smiol.c]: GNU is warning that `ierr` is unused in this function (and in SMIOL_close_file) when `PNETCDF` is not defined. Is it necessary to use `ierr` in this conditional? We could do the following and achive the same results without `ierr`.

``` C
if (ncmpi_create(MPI_Comm_f2c(context->fcomm), filename, NC_NOCLOBBER, MPI_INFO_NULL, &((*file)->ncidp)) != NC_NOERR)
```

File: src/smiol.c [hunk 4]
```diff
@@ -135,11 +139,51 @@ int SMIOL_inquire(void)
...
+		return SMIOL_MALLOC_FAILURE;
+	}
+
+	/*
+	 * Save pointer to context for this file
+	 */
+	(*file)->context = context;
+
+#ifdef SMIOL_PNETCDF
+	if ((ierr = ncmpi_create(MPI_Comm_f2c(context->fcomm), filename, NC_NOCLOBBER, MPI_INFO_NULL, &((*file)->ncidp))) != NC_NOERR) {
```
Comment (MiCurry) [inline src/smiol.c]: This line is also longer then 80 characters. Should it be broken into more lines?

**Comment by mgduda** [comment]
@MiCurry I think the segmentation faults may be due to a failure to create an existing netCDF file. In `SMIOL_open_file`, the mode for the file to be created is `NC_NOCLOBBER`. We can change this in a future PR, and we don't yet have a way to get library-specific error codes and messages (though I have a branch that does handle these...). Can you try removing all `*.nc` files before running the `smiol_runner_c` and `smiol_runner_f` programs?

**Comment by MiCurry** [comment]
> @MiCurry I think the segmentation faults may be due to a failure to create an existing netCDF file. In `SMIOL_open_file`, the mode for the file to be created is `NC_NOCLOBBER`. We can change this in a future PR, and we don't yet have a way to get library-specific error codes and messages (though I have a branch that does handle these...). Can you try removing all `*.nc` files before running the `smiol_runner_c` and `smiol_runner_f` programs?

Yes, I just discovered that! The segmentation fault was caused because the files that .nc had already been created.

**Comment by mgduda** [comment]
@MiCurry I've just fixed the capitalization of `SMIOL_success`, and I've removed the used `ierr` variable.

**Pull request merged**